extensions/stackdriver/edges/edge_reporter.h

```diff
@@ -0,0 +1,100 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "extensions/common/context.h"
+#include "extensions/stackdriver/edges/edges.pb.h"
+#include "extensions/stackdriver/edges/mesh_edges_service_client.h"
+#include "google/protobuf/util/time_util.h"
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Edges {
+
+#ifdef NULL_PLUGIN
+using Envoy::Extensions::Common::Wasm::Null::Plugin::Extensions::Stackdriver::
+    Edges::MeshEdgesServiceClient;
+#endif
+
+using Envoy::Extensions::Common::Wasm::Null::Plugin::getCurrentTimeNanoseconds;
+using google::cloud::meshtelemetry::v1alpha1::ReportTrafficAssertionsRequest;
+using google::cloud::meshtelemetry::v1alpha1::WorkloadInstance;
+using google::protobuf::util::TimeUtil;
+
+namespace {
+static inline google::protobuf::Timestamp wasmTimestamp() {
+  return TimeUtil::NanosecondsToTimestamp(getCurrentTimeNanoseconds());
+}
+}  // namespace
+
+// EdgeReporter provides a mechanism for generating information on traffic
+// "edges" for a mesh. It should be used **only** to document incoming edges for
+// a proxy. This means that the proxy in which this reporter is running should
+// be the destination workload instance for all reported traffic.
+// This should only be used in a single-threaded context. No support for
+// threading is currently provided.
+class EdgeReporter {
+  typedef std::function<google::protobuf::Timestamp()> TimestampFn;
+
+ public:
+  EdgeReporter(const ::wasm::common::NodeInfo &local_node_info,
```

Comment:
I think you need a destructor that flushes/reports the edges. You might need to be careful with the lifetime of the client as the reporter is being destructed (e.g. callbacks might become stale).

Reply:
good point, I need to add this to logging part as well..

Reply:
Added a simple destructor which should call `reportEdges()`. That should be OK re: lifetimes I believe, as the client should be destructed after completion of the destructor, correct?

Reply:
Yes it should be fine, it need special care when we add retry in the future.